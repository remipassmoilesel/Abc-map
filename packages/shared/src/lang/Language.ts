/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * Supported languages
 *
 * Values must be compatible with lang attribute: https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/lang
 */
export enum Language {
  English = 'en',
  French = 'fr',
}

export const SupportedLanguages = Object.values(Language);

export const FallbackLang = Language.English;

export function isLangSupported(lang: string | Language): boolean {
  return Object.values(Language).includes(lang as Language);
}

export function langFromString(lang: string): Language | undefined {
  return Object.values(Language).find((l) => l === lang);
}
