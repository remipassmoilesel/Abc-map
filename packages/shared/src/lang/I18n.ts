/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { FallbackLang, Language } from './Language';

export interface I18nText {
  language: Language;
  text: string;
}

export interface I18nList {
  language: Language;
  text: string[];
}

export function getTextByLang(texts: I18nText[], lang: Language | string): string | undefined {
  const wanted = texts.find((desc) => desc.language === lang);
  const fallback = texts.find((desc) => desc.language === FallbackLang);
  const first = texts.length ? texts[0] : undefined;
  return wanted?.text ?? fallback?.text ?? first?.text;
}

export function getListByLang(texts: I18nList[], lang: Language | string): string[] | undefined {
  const wanted = texts.find((desc) => desc.language === lang);
  const fallback = texts.find((desc) => desc.language === FallbackLang);
  const first = texts.length ? texts[0] : undefined;
  return wanted?.text ?? fallback?.text ?? first?.text;
}
