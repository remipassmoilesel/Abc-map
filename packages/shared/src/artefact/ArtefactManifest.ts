/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { I18nList, I18nText } from '../lang';
import { ArtefactType } from './ArtefactType';

/**
 * This interface represents an artefact yaml manifest file.
 *
 * An artefact is a bundle of data or a pre-configured data source, usable in Abc-Map.
 *
 */
export interface ArtefactManifest {
  /**
   * Format version, 0.0.1 for the moment
   */
  version: string;
  artefact: {
    /**
     * Name of artefact, displayed in lists
     */
    name: I18nText[];
    /**
     * Type: vector or basemap
     */
    type: ArtefactType;
    /**
     * Full description. Can be HTML. You do not need to create HTML links, they will be created automatically.
     */
    description?: I18nText[];
    /**
     * Keywords attached to artefact. They will be used in search.
     */
    keywords?: I18nList[];
    /**
     * Full text license or list of links.
     */
    license: string;
    /**
     * Name of the entity which provide or produce data.
     */
    provider: string;
    /**
     * Link to the source of the data.
     */
    link: string;
    /**
     * Attributions that will be displayed on maps
     */
    attributions: string[];
    /**
     * List of paths to artefact files. Relative to this manifest.
     */
    files: string[];
    /**
     * List of paths to preview images. Relative to this manifest.
     *
     * Can be all format supported by <img/> tag.
     */
    previews?: string[];
    /**
     * Weight is used to sort results in search and lists. The higher the number, the more likely the artifact is above.
     */
    weight?: number;

    metadata?: Record<string, unknown>;
  };
}
