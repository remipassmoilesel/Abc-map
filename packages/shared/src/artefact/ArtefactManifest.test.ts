/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */
import { Language } from '../lang';
import { ArtefactType } from './ArtefactType';
import { ArtefactManifest } from './ArtefactManifest';

describe('ArtefactManifest', () => {
  it('should not change without datastore migration', () => {
    const manifest: ArtefactManifest = {
      version: 'version',
      artefact: {
        name: [
          { language: Language.French, text: 'name' },
          { language: Language.English, text: 'name' },
        ],
        type: ArtefactType.BaseMap,
        provider: 'provider',
        description: [
          { language: Language.French, text: 'description' },
          { language: Language.English, text: 'description' },
        ],
        keywords: [
          { language: Language.French, text: ['keywords'] },
          { language: Language.English, text: ['keywords'] },
        ],
        license: 'license',
        attributions: ['attributions'],
        link: 'link',
        files: ['file'],
        previews: ['previews'],
      },
    };

    expect(manifest).toMatchSnapshot();
  });
});
