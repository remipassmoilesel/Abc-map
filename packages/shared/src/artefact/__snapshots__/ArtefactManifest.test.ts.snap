// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtefactManifest should not change without datastore migration 1`] = `
Object {
  "artefact": Object {
    "attributions": Array [
      "attributions",
    ],
    "description": Array [
      Object {
        "language": "fr",
        "text": "description",
      },
      Object {
        "language": "en",
        "text": "description",
      },
    ],
    "files": Array [
      "file",
    ],
    "keywords": Array [
      Object {
        "language": "fr",
        "text": Array [
          "keywords",
        ],
      },
      Object {
        "language": "en",
        "text": Array [
          "keywords",
        ],
      },
    ],
    "license": "license",
    "link": "link",
    "name": Array [
      Object {
        "language": "fr",
        "text": "name",
      },
      Object {
        "language": "en",
        "text": "name",
      },
    ],
    "previews": Array [
      "previews",
    ],
    "provider": "provider",
    "type": "basemap",
  },
  "version": "version",
}
`;
